### DISCUSSION TITLE
Secret leak prevention out of the box

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/next.js#33386] [Ideas]
Hi guys, do you think it would make sense to include a secret leak check out-of-the-box in Next.js? 

I've tried Gitleaks (https://github.com/VulcanJS/vulcan-next/issues/143), I am not yet 100% confident in it. I wonder if smth similar could be applied specifically to Next.js projects? Or maybe simply an official documentation on the subject? 
The difficulty also stems for the dependency to either Go, or Docker to avoid the install, so it takes a few second to run just because of the container.

It seems too easy to me to accidentally use a .env.local with some sensible information, eg for temporary testing, accidentaly destroying .gitignore or accidentally rename it, and push it to the repo, or even accidentally adding variables for some reason. In my scenario, I temporarily copied a file with env variables in order to get the exact name of a single value, but forgot to remove the pasted code, which also contained a sensitive value... 

I see there as been discussion about this recently but focused on a weird use case (https://github.com/vercel/next.js/discussions/32874), here I am talking about actual developer error, that could be gracefully caught by the framework. 

One related scenario I also dread is the accidental leak of getServerSideProps into the client bundle, which could also lead to env variables leaking but at runtime instead. I know there are unit tests and solid code to handle this, but I'd sleep better with a second layer of security.

I know this isn't specific to Next.js, but maybe your Rust superpowers could help designing a nice solution :)

**Comment by theoephraim** [comment]
Not built-in, but [DMNO's Next.js plugin](https://dmno.dev/docs/integrations/nextjs/) scans bundled code and outgoing server responses for your actual secret values.

**Comment by theoephraim** [comment]
As does https://varlock.dev - which is a newer iteration of DMNO but is simpler.

**Comment by theoephraim** [comment]
Latest [varlock nextjs integration](https://varlock.dev/integrations/nextjs/) redacts secrets in logs, and prevents leaked secrets at build and runtime -- check it out!

**Comment by Hybirdss** [comment]
This thread has been open 4 years with community interest but no maintainer verdict on whether a build-time warning approach is welcome. Rather than submit a cold PR and risk close-as-duplicate-discussion, leaving a concrete proposal here for feedback.

## Proposal

A build-time warn-once check in `getNextPublicEnvironmentVariables()` (`packages/next/src/lib/static-env.ts`) that fires `Log.warnOnce` when a `NEXT_PUBLIC_*` value matches one of a curated list of anchored high-confidence credential shapes. The prefix semantics (`NEXT_PUBLIC_` = intentionally public) are preserved — the warning never blocks build, never alters emitted code, and is opt-out-able via `NEXT_SUPPRESS_PUBLIC_ENV_SHAPE_WARNING=1`.

Design choices made to minimize false-positive churn (the #1 reason the framework team has historically paused on this class of change):

- **Anchored regex only**, no entropy heuristics. 16 patterns cover Tier-1 providers: Stripe `sk_(live|test)_`, AWS `AKIA`, GitHub `ghp_`/`gho_`/`github_pat_`/`ghs_`, OpenAI `sk-(proj|svcacct|admin)-` and legacy `T3BlbkFJ`, Anthropic `sk-ant-api03-`, Slack `xoxb-`/`xoxp-`, SendGrid, npm, PEM private-key headers, Supabase `service_role` JWT. Every regex is anchored (`^...$`) and requires a minimum tail length.

- **Public-by-design identifiers are explicitly excluded** from the warn list: Firebase `apiKey`, Stripe publishable key `pk_`, PostHog `phc_`, Supabase anon JWT, Google Maps `AIzaSy`, Algolia search key, Sentry DSN, Mapbox `pk.`. These all flow through `NEXT_PUBLIC_*` legitimately and would never trigger a false warning.

- **16-character minimum** avoids sentinels (`true`, `development`, `1`) ever matching.

- **`warnOnce`** means no noisy repeat during hot-reload.

- **No behavior change**: if a user intentionally ships a `NEXT_PUBLIC_API_TOKEN` that happens to look like a Stripe `sk_live_`, they get one warning and can either rename or set `NEXT_SUPPRESS_PUBLIC_ENV_SHAPE_WARNING=1`.

## What it looks like

```
warn  - Next.js detected a NEXT_PUBLIC_* env var whose value matches a known
        credential shape (Stripe secret key).

        NEXT_PUBLIC_STRIPE_SECRET=sk_live_***

        NEXT_PUBLIC_ values are inlined into the browser bundle and shipped to
        every user. If this is intentional, set
        NEXT_SUPPRESS_PUBLIC_ENV_SHAPE_WARNING=1. Otherwise rename it to omit
        NEXT_PUBLIC_.
```

## Prior art

No framework in the ecosystem currently ships this class of warning — Vite, SvelteKit, CRA all inline their public-prefix env vars without value inspection. Next.js shipping it first would establish a reasonable default.

## Branch + patch

I have the three-file patch (`check-public-env-shape.ts` new, `static-env.ts` 2-line edit, 170-line unit test) ready. It applies cleanly against current `canary` HEAD and uses existing `Log.warnOnce` infrastructure.

Two questions before opening a PR:

1. Is the build-time warn-once approach acceptable in principle to the team? (I don't want to submit cold and duplicate this thread.)
2. If yes — preferred integration point: `getNextPublicEnvironmentVariables` directly, or factored into the `@next/eslint-plugin-next` as a new rule that the framework default-enables?

Happy to reshape either way. Will hold on PR until a maintainer weighs in.

(Context, for transparency: this proposal was sharpened after the April 2026 Vercel incident publicly highlighted how much friction exists between "`NEXT_PUBLIC_` was designed to be public" and "customers are misusing it as a shorthand for any env var." The framework warning is a zero-blast-radius hardening; no exploits are being withheld from a PR by keeping the proposal public here.)

**Reply by eps1lon** [reply]
The nature of a global opt-out feels too coarse to me. You add PAT for a burner account, disable the warning, and now you're back to square one when you accidentally add a new, sensitive secret later on. An opt-out per variable would make more sense but isn't that already what the `NEXT_PUBLIC` prefix does?

Keep in mind that it's not easy to include a sensitive secret in browser bundles. You actually have to prefix it with `NEXT_PUBLIC_`. E.g. `process.env.GITHUB_TOKEN` will **not** be filled in during build even if there is a `GITHUB_TOKEN` available. Only `process.env.NEXT_PUBLIC_GITHUB_TOKEN` would work if `NEXT_PUBLIC_GITHUB_TOKEN` is available during build time. See https://nextjs.org/docs/pages/guides/environment-variables#bundling-environment-variables-for-the-browser

**Reply by Hybirdss** [reply]
Thanks for the feedback — both points are well taken.

**On the global opt-out**: Agreed — dropped. I should also acknowledge the gap this creates: without any escape hatch, the only documented silence path is renaming the variable. If the team identifies a concrete use case that needs per-variable suppression, I think a `next.config.js` allowlist is the right shape (rather than another env var):

```js
// next.config.js
experimental: {
  suppressPublicEnvShapeWarnings: ['NEXT_PUBLIC_MY_KEY']
}
```

Happy to add that in a follow-up or in this PR if it's a blocker. For the current Tier-1 set specifically, I don't know of a case where `sk_live_`, `AKIA`, or `ghp_` is legitimately public — the patterns were chosen to exclude all public-by-design identifiers — but I could be missing something.

**On "isn't the prefix already the opt-in"**: That's a fair framing. The distinction I had in mind is that the prefix opts-in the variable *name*, while this checks the *value content* — but I recognize that might not be a distinction worth encoding in the framework. If the team's read is that this belongs in a linter or external scanner (gitleaks, trufflehog) instead, I think that's reasonable. Happy to close this PR and take it that direction.

Either way, appreciate you taking a look.

**Comment by Hybirdss** [comment]
Follow-up: I've opened #93005 as a **draft** containing the 3-file reference implementation described above. The draft status is intentional — it's there as concrete code to look at during this thread's discussion, not a request for review.

Will flip from draft to ready-for-review only after a maintainer weighs in on the approach here. Happy to close the draft entirely if the team's direction differs.
